### PR TITLE
drivers/video: Add config for API compatibility

### PR DESCRIPTION
**Pull request by SPRESENSE**
## Summary

Add a Kconfig option for supporting old APIs on video driver.
The video_initialize() and video_uninitialize() are re-named to capture_initialize() and capture_uninitialize().
But this modification made break compatibility on user APIs.
So this commit add a option on Kconfig to use old API symbol name.

## Impact

No impact except enabling the additional Kconfig option.
Video I/F has an impact when the option is enabled.

## Testing

Test spresense:camera, if it can be build and works well.

### CODE REVIEW

**Comment by acassis** [comment]
@SPRESENSE isn't it better to fix the existing applications using the old API instead bringing back those APIs to mainline?

I agree that breaking APIs is not a good thing, but in this case the previous change seems correct: for video capture it is better to use capture_initialize() than video_initialize().

We need to reduce the Kconfig complexity, not increasing it. This patch is increasing complexity for something that should be fixed in the right way (in the external program using this kernel).

File: drivers/video/Kconfig [hunk 1]
```diff
@@ -163,6 +163,14 @@ config VIDEO_SCENE_TEXT
 	default y
 	---help---
 		Enable text scene
+
+config VIDEO_COMPAT_OLD_APIS
```
Comment (xiaoxiang781216) [inline drivers/video/Kconfig:167]: let's remove the option and define macro always if the combability is required.

**Comment by xiaoxiang781216** [comment]
But these APIs shouldn't be called from userspace, so we have some flexibility to modify them. BTW, @shizhenghui made this change because v4l2 m2m is added recently, it isn't good to name capture_xxx as video_xxx.

**Review by acassis — CHANGES_REQUESTED** [review]
Avoid creating Kconfig entrypoint and just create the macro and include some info that it will be deprecated and the developers should use capture_initialize()

**Comment by hartmannathan** [comment]
> Avoid creating Kconfig entrypoint and just create the macro and include some info that it will be deprecated and the developers should use capture_initialize()

Agreed, but how do we make a noticeable warning (at build-time) that the deprecated API is used? It would be best if messages could be printed at end of compile output.

**Comment by acassis** [comment]
> > Avoid creating Kconfig entrypoint and just create the macro and include some info that it will be deprecated and the developers should use capture_initialize()
> 
> Agreed, but how do we make a noticeable warning (at build-time) that the deprecated API is used? It would be best if messages could be printed at end of compile output.

There is a deprecated attribute, it is used on Linux kernel.

We could create it for NuttX as well:

```
#ifdef __GNUC__
#define __deprecated __attribute__((deprecated))
#else
#define __deprecated
#endif
```

**Comment by hartmannathan** [comment]
The build on msys2 failed. I restarted the job with debug logging. Let's see if the error is in the codebase or just a spurious error...

**Comment by hartmannathan** [comment]
> > > Avoid creating Kconfig entrypoint and just create the macro and include some info that it will be deprecated and the developers should use capture_initialize()
> > 
> > 
> > Agreed, but how do we make a noticeable warning (at build-time) that the deprecated API is used? It would be best if messages could be printed at end of compile output.
> 
> There is a deprecated attribute, it is used on Linux kernel.
> 
> We could create it for NuttX as well:
> 
> ```
> #ifdef __GNUC__
> #define __deprecated __attribute__((deprecated))
> #else
> #define __deprecated
> #endif
> ```
> 
> So we just add:
> 
> ```
> void old_function(void) __deprecated;
> 
> void old_function(void)
> {
>     // implementation
> }
> ```

@acassis thank you!

I looked up attribute __deprecated and found this StackOverflow answer which covers GCC, Clang, and MSVC:

https://stackoverflow.com/a/21265197

tl;dr GCC and Clang use __attribute__((deprecated)) (but we should check for __GNUC__ || __clang__) while MSVC uses __declspec(deprecated).

**Comment by acassis** [comment]
> > > > Avoid creating Kconfig entrypoint and just create the macro and include some info that it will be deprecated and the developers should use capture_initialize()
> > > 
> > > 
> > > Agreed, but how do we make a noticeable warning (at build-time) that the deprecated API is used? It would be best if messages could be printed at end of compile output.
> > 
> > 
> > There is a deprecated attribute, it is used on Linux kernel.
> > We could create it for NuttX as well:
> > ```
> > #ifdef __GNUC__
> > #define __deprecated __attribute__((deprecated))
> > #else
> > #define __deprecated
> > #endif
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > So we just add:
> > ```
> > void old_function(void) __deprecated;
> > 
> > void old_function(void)
> > {
> >     // implementation
> > }
> > ```
> 
> @acassis thank you!
> 
> I looked up attribute __deprecated and found this StackOverflow answer which covers GCC, Clang, and MSVC:
> 
> https://stackoverflow.com/a/21265197
> 
> tl;dr GCC and Clang use **attribute**((deprecated)) (but we should check for **GNUC** || **clang**) while MSVC uses __declspec(deprecated).

Yes, but I think we don't need to implement MSVC since we don't have support for it.

**Comment by xiaoxiang781216** [comment]
> The build on msys2 failed. I restarted the job with debug logging. Let's see if the error is in the codebase or just a spurious error...

the patch need to rebase to pass ci since the fix is provided after this pr: https://github.com/apache/nuttx/pull/12788

**Comment by SPRESENSE** [comment]
Hi,
Thank you for your comments.

I agree that this functions should not be called by userspace.
In nuttx, there is a function board_ioctl() to control board specific devices.

So I think we should define a special command for initializing video device and use it insted.

I will make a patch and force push it on this PR.
And also rebase to newest main branch.
